oscore-cf/src/main/java/org/eclipse/californium/oscore/OSSerializer.java

```
@@ -0,0 +1,557 @@
+	 * 
+	 *
+	 * @return the serialized AAD for OSCore
+	 */
+	public static byte[] serializeReceiveRequestAAD(int version, int seq, OSCoreCtx ctx, OptionSet options) {
+		if (version == CoAP.VERSION) {
+			if (seq > -1) {
+				if (ctx != null) {
+					if (options != null) {
+						CBORObject algorithms = CBORObject.NewArray();
+						algorithms.Add(ctx.getAlg().AsCBOR());
+
+						CBORObject aad = CBORObject.NewArray();
+						aad.Add(version);
+						aad.Add(algorithms);
+						aad.Add(ctx.getRecipientId());
+						aad.Add(processPartialIV(seq));
+						return aad.EncodeToBytes();
+					} else {
+						LOGGER.severe(OSExceptionHandler.OPTIONSET_NULL);
+						throw new IllegalArgumentException(OSExceptionHandler.OPTIONSET_NULL);
+					}
+				} else {
+					LOGGER.severe(OSExceptionHandler.CTX_NULL);
+					throw new IllegalArgumentException(OSExceptionHandler.CTX_NULL);
+				}
+			} else {
+				LOGGER.severe(OSExceptionHandler.SEQ_NBR_INVALID);
+				throw new IllegalArgumentException(OSExceptionHandler.SEQ_NBR_INVALID);
+			}
+		} else {
+			LOGGER.severe(OSExceptionHandler.WRONG_VERSION_NBR);
+			throw new IllegalArgumentException(OSExceptionHandler.WRONG_VERSION_NBR);
+		}
+	}
+
+	/**
+	 * Prepare the E or I options for encrypting.
+	 * 
+	 * @param writer the datagram writer collecting the necessary bytes
+	 * @param optionSet the original options
+	 * @param isConfidential true will use E options since we are serializing confidential data, false will use I options
+	 * @param request true is request processing, false is response processing
+	 * @param realCode the real code
+	 */
+	private static void writeOptions(DatagramWriter writer, OptionSet optionSet, boolean isConfidential, boolean request,
+			int realCode) {
+		int previousOptionNumber = 0;
+
+		OptionSet filteredOptions = null;
+		if (isConfidential) {
+			filteredOptions = OptionJuggle.prepareEoptions(optionSet);
+		}
+
+		if (realCode > 0) {
+			writer.writeBytes(new byte[] { (byte) realCode });
```

Are you sure? 0 is not written? Or 0 should thrown an Exception?

---

I should add a general check against the possible CoAP codes (1,2,3,4,5, 6). Yes, I should throw an exception.